### PR TITLE
Changed read_csv type hint for na_values and added tests Issue#53813

### LINKED ISSUE #53813: BUG: `read_csv` type hint for `na_values` should include non-iterables as valid arguments
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
df = pd.DataFrame({'a': [1,2,3], 'b': list('hij'), 'c': ['A', 'N/A', 'BBB']})
df.to_csv('test.csv', index=False)

# passing scalar argument
pd.read_csv('test.csv', na_values=2)
# output:
#       a  b    c
#  0  1.0  h    A
#  1  NaN  i  NaN
#  2  3.0  j  BBB

# passing string argument
pd.read_csv('test.csv', na_values='BBB')
# output:
#     a  b    c
#  0  1  h    A
#  1  2  i  NaN
#  2  3  j  NaN

# passing list of scalar and string as argument
>>> pd.read_csv('test.csv', na_values=[2, 'BBB'])
#       a  b    c
#  0  1.0  h    A
#  1  NaN  i  NaN
#  2  3.0  j  NaN
```

### Issue Description

The description of valid types in the `read_csv` docstring under `na_values` shows:
> *scalar, str, list-like, or dict, optional*

However, the type signature implies that only list-like or dict are valid (not scalar or str):
> `Sequence[str] | Mapping[str, Sequence[str]] | None = None`

It appears that the function does actually accept scalar or str (i.e. non-list-list, non-iterable) values as evidenced by the example above, meaning the description is accurate and the type hint needs to be updated to match. The main inconsistencies with the type hint are the following:
* No indication of scalar or str being a valid argument
* The sequence of str is too specific since a list containing str and scalar appears to be valid (e.g., `[2, 'BBB']` seems to work).
* The mapping also appears too specific, since the keys correspond to column names which are not always str (could be any Hashable) and the values in the mapping are too specific for the same reasons listed in the previous bullet. 

Note: this was one of a few items I mentioned in #53763 but I thought it might need its own issue since it affects more than just the docstring. If it should have still been labelled 'DOC', please feel free to update.

### Expected Behavior

The behavior seems to be correct, but the type hint should match, perhaps by updating to something like this:

`Hashable | Iterable[Hashable] | Mapping[Hashable, Iterable[Hashable]] | None = None` (assuming `Hashable` is the correct type to cover 'scalar or str` as shown in the description but perhaps someone more familiar could confirm this).

### Installed Versions

<details>
/home/tpaxman/mambaforge/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")

INSTALLED VERSIONS
------------------
commit           : 965ceca9fd796940050d6fc817707bba1c4f9bff
python           : 3.10.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.15.90.1-microsoft-standard-WSL2
Version          : #1 SMP Fri Jan 27 02:56:13 UTC 2023
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.2
numpy            : 1.25.0
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 65.6.3
pip              : 23.0.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           :
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : 0.19.0
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by kvn4**
- [ ] closes #53813
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.

### CODE REVIEW

File: pandas/tests/io/parser/test_na_values.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 xfail_pyarrow = pytest.mark.usefixtures("pyarrow_xfail")
 
 
+@skip_pyarrow
```
Comment (kvn4) [inline pandas/tests/io/parser/test_na_values.py:23]: The pyarrow compiler raised a "TypeError: expected bytes, ellipsis found". I noticed some of the other tests had @skip_pyarrow to bypass this which is why I added it to the tests that were getting this type error.

Screen Shot 2023-08-10 at 12 35 34 PM

Comment (lithomas1) [reply pandas/tests/io/parser/test_na_values.py:23]: The ellipsis is incorrect here I think. You are changing the default value of the function in addition to the type hint.

File: pandas/tests/io/parser/test_na_values.py [hunk 2]
```diff
@@ -669,3 +676,47 @@ def test_bool_and_nan_to_float(all_parsers):
     result = parser.read_csv(StringIO(data), dtype="float")
     expected = DataFrame.from_dict({"0": [np.nan, 1.0, 0.0]})
     tm.assert_frame_equal(result, expected)
+
+
+@skip_pyarrow
+def test_boolean_na_values(all_parsers):
```
Comment (twoertwein) [inline pandas/tests/io/parser/test_na_values.py:682]: Pandas does typically not add tests for type annotations (pandas-stubs does https://github.com/pandas-dev/pandas-stubs/). Most of the pandas tests are not checked by mypy (and pyright).

File: pandas/io/parsers/readers.py [hunk 3]
```diff
@@ -637,7 +638,10 @@ def read_csv(
     skiprows: list[int] | int | Callable[[Hashable], bool] | None = ...,
     skipfooter: int = ...,
     nrows: int | None = ...,
-    na_values: Sequence[str] | Mapping[str, Sequence[str]] | None = ...,
+    na_values: Hashable
+    | Iterable[Hashable]
+    | Mapping[Hashable, Iterable[Hashable]]
+    | None = ...,
```
Comment (twoertwein) [inline pandas/io/parsers/readers.py:644]: You could remove `None` if you like to, it is `Hashable`.

File: pandas/io/parsers/readers.py [hunk 4]
```diff
@@ -878,7 +891,10 @@ def read_csv(
     skipfooter: int = 0,
     nrows: int | None = None,
     # NA and Missing Data Handling
-    na_values: Sequence[str] | Mapping[str, Sequence[str]] | None = None,
+    na_values: Hashable
+    | Iterable[Hashable]
+    | Mapping[Hashable, Iterable[Hashable]]
+    | None = ...,
```
Comment (twoertwein) [inline pandas/io/parsers/readers.py:897]: `= None`, `...` is only used in the overloads

**Comment by kvn4** [comment]
Hi @twoertwein and @lithomas1 thank you for the feedback. Do you know where the ellipses issue could be coming from? I only changed the na_values type hint for the read_csv() function.

**Comment by twoertwein** [comment]
> Hi @twoertwein and @lithomas1 thank you for the feedback. Do you know where the ellipses issue could be coming from? I only changed the na_values type hint for the read_csv() function.

You seem to have changed it :) You can always use `git add --patch` to confirm individual changes you want to add.

You can simply replace `= ...` with `= None` in the non-overload definition.

The easiest way to remove the test is probably `git checkout <commit id from before your changes> pandas/tests/io/parser/test_na_values.py` and then add and commit that.

**Comment by kvn4** [comment]
Hi @twoertwein do the changes I made look good?

**Comment by twoertwein** [comment]
> Hi @twoertwein do the changes I made look good?

Now you use `= None` everywhere. You want `= None` for the function definition that doesn't have the `@overload` but otherwise use `= ...`.

It might be better to use `Sequence` instead of `Iterable` (not sure whether any iterable is accepted).

You can remove the test changes you made.

**Comment by kvn4** [comment]
> > Hi @twoertwein do the changes I made look good?
> 
> Now you use `= None` everywhere. You want `= None` for the function definition that doesn't have the `@overload` but otherwise use `= ...`.
> 
> It might be better to use `Sequence` instead of `Iterable` (not sure whether any iterable is accepted).
> 
> You can remove the test changes you made.

Thank you for the feedback. I made the changes

**Review by twoertwein — APPROVED** [review]

**Comment by twoertwein** [comment]
Thanks @kvn4 !

**Pull request merged**